In [271]:
import pandas as pd
import numpy as np
import time

In [208]:
train_data=pd.read_csv("/Users/gardasnagarjun/Downloads/Assignment2_v3/Q1_dataset/letter_train",sep=" ",header=None)
train_data.columns=['Alphabet','feat1','feat2','feat3','feat4','feat5','feat6','feat7','feat8','feat9','feat10','feat11','feat12','feat13','feat14','feat15','feat16','feat17']
train_data.head()

Alphabet         feat1         feat2        feat3         feat4  \
0        26   1:-0.733333   2:-0.466667  3:-0.466667        4:-0.6   
1        16   1:-0.466667  2:-0.0666667  3:-0.333333   4:-0.333333   
2        19  1:-0.0666667    2:0.333333  3:0.0666667  4:-0.0666667   
3         8   1:-0.466667         2:0.2  3:-0.333333  4:-0.0666667   
4         8        1:-0.2  2:-0.0666667  3:0.0666667   4:-0.333333   

         feat5         feat6         feat7        feat8         feat9  \
0  5:-0.733333  6:-0.0666667   7:0.0666667  8:-0.733333         9:0.2   
1  5:-0.333333   6:-0.333333         7:0.2       8:-0.2   9:-0.466667   
2  5:-0.466667   6:0.0666667   7:0.0666667  8:-0.333333    9:0.333333   
3  5:-0.466667  6:-0.0666667  7:-0.0666667   8:0.733333   9:-0.866667   
4  5:-0.466667  6:-0.0666667        7:-0.2       8:-0.6  9:-0.0666667   

          feat10         feat11         feat12        feat13        feat14  \
0    10:0.466667  11:-0.0666667  12:-0.0666667  13:-0.866667  14:0.0666667   
1   10:0.0666667  11:-0.0666667         12:0.2  13:-0.733333        14:0.2   
2    10:0.466667   11:-0.733333   12:0.0666667  13:-0.733333  14:-0.333333   
3  10:-0.0666667        11:-0.2   12:0.0666667       13:-0.6  14:0.0666667   
4    10:0.333333  11:-0.0666667         12:0.2       13:-0.6  14:0.0666667   

          feat15         feat16  feat17  
0   15:-0.333333        16:-0.2     NaN  
1  15:-0.0666667    16:0.333333     NaN  
2   15:-0.333333    16:0.333333     NaN  
3          15:-1   16:0.0666667     NaN  
4        15:-0.6  16:-0.0666667     NaN

In [209]:
train_data.head()

Alphabet         feat1         feat2        feat3         feat4  \
0        26   1:-0.733333   2:-0.466667  3:-0.466667        4:-0.6   
1        16   1:-0.466667  2:-0.0666667  3:-0.333333   4:-0.333333   
2        19  1:-0.0666667    2:0.333333  3:0.0666667  4:-0.0666667   
3         8   1:-0.466667         2:0.2  3:-0.333333  4:-0.0666667   
4         8        1:-0.2  2:-0.0666667  3:0.0666667   4:-0.333333   

         feat5         feat6         feat7        feat8         feat9  \
0  5:-0.733333  6:-0.0666667   7:0.0666667  8:-0.733333         9:0.2   
1  5:-0.333333   6:-0.333333         7:0.2       8:-0.2   9:-0.466667   
2  5:-0.466667   6:0.0666667   7:0.0666667  8:-0.333333    9:0.333333   
3  5:-0.466667  6:-0.0666667  7:-0.0666667   8:0.733333   9:-0.866667   
4  5:-0.466667  6:-0.0666667        7:-0.2       8:-0.6  9:-0.0666667   

          feat10         feat11         feat12        feat13        feat14  \
0    10:0.466667  11:-0.0666667  12:-0.0666667  13:-0.866667  14:0.0666667   
1   10:0.0666667  11:-0.0666667         12:0.2  13:-0.733333        14:0.2   
2    10:0.466667   11:-0.733333   12:0.0666667  13:-0.733333  14:-0.333333   
3  10:-0.0666667        11:-0.2   12:0.0666667       13:-0.6  14:0.0666667   
4    10:0.333333  11:-0.0666667         12:0.2       13:-0.6  14:0.0666667   

          feat15         feat16  feat17  
0   15:-0.333333        16:-0.2     NaN  
1  15:-0.0666667    16:0.333333     NaN  
2   15:-0.333333    16:0.333333     NaN  
3          15:-1   16:0.0666667     NaN  
4        15:-0.6  16:-0.0666667     NaN

In [210]:
columns=['feat1','feat2','feat3','feat4','feat5','feat6','feat7','feat8','feat9','feat10','feat11','feat12','feat13','feat14','feat15','feat16']
# Removing the first two characters in every data attribute value in the training dataset
for c in columns:
    train_data[c] = train_data[c].str.split(':').str[1]
train_data.head()

Alphabet       feat1       feat2      feat3       feat4      feat5  \
0        26   -0.733333   -0.466667  -0.466667        -0.6  -0.733333   
1        16   -0.466667  -0.0666667  -0.333333   -0.333333  -0.333333   
2        19  -0.0666667    0.333333  0.0666667  -0.0666667  -0.466667   
3         8   -0.466667         0.2  -0.333333  -0.0666667  -0.466667   
4         8        -0.2  -0.0666667  0.0666667   -0.333333  -0.466667   

        feat6       feat7      feat8       feat9      feat10      feat11  \
0  -0.0666667   0.0666667  -0.733333         0.2    0.466667  -0.0666667   
1   -0.333333         0.2       -0.2   -0.466667   0.0666667  -0.0666667   
2   0.0666667   0.0666667  -0.333333    0.333333    0.466667   -0.733333   
3  -0.0666667  -0.0666667   0.733333   -0.866667  -0.0666667        -0.2   
4  -0.0666667        -0.2       -0.6  -0.0666667    0.333333  -0.0666667   

       feat12     feat13     feat14      feat15      feat16  feat17  
0  -0.0666667  -0.866667  0.0666667   -0.333333        -0.2     NaN  
1         0.2  -0.733333        0.2  -0.0666667    0.333333     NaN  
2   0.0666667  -0.733333  -0.333333   -0.333333    0.333333     NaN  
3   0.0666667       -0.6  0.0666667          -1   0.0666667     NaN  
4         0.2       -0.6  0.0666667        -0.6  -0.0666667     NaN

In [211]:
test_data=pd.read_csv("/Users/gardasnagarjun/Downloads/Assignment2_v3/Q1_dataset/letter_test",sep=" ",header=None)
test_data.columns=['Alphabet','feat1','feat2','feat3','feat4','feat5','feat6','feat7','feat8','feat9','feat10','feat11','feat12','feat13','feat14','feat15','feat16','feat17']
test_data.head()

Alphabet        feat1         feat2        feat3        feat4        feat5  \
0        17  1:-0.466667  2:-0.0666667  3:-0.333333       4:-0.2  5:-0.733333   
1        18       1:-0.6         2:0.2  3:-0.466667       4:-0.2       5:-0.6   
2         6  1:-0.733333   2:-0.866667  3:-0.733333  4:-0.733333  5:-0.866667   
3         6  1:-0.733333   2:-0.466667  3:-0.466667       4:-0.6  5:-0.733333   
4        20  1:-0.333333   2:0.0666667       3:-0.2       4:-0.2       5:-0.6   

          feat6       feat7        feat8         feat9         feat10  \
0   6:0.0666667      7:-0.2  8:0.0666667        9:-0.2        10:-0.2   
1   6:-0.333333  7:0.333333        8:0.2        9:-0.6  10:-0.0666667   
2   6:-0.333333  7:0.333333  8:-0.466667   9:-0.333333    10:0.333333   
3  6:-0.0666667       7:0.2  8:-0.866667  9:-0.0666667    10:0.733333   
4   6:-0.466667  7:0.733333  8:-0.466667        9:-0.2         10:0.6   

         feat11        feat12        feat13         feat14        feat15  \
0       11:-0.2  12:0.0666667       13:-0.6   14:0.0666667  15:-0.466667   
1  11:-0.466667  12:0.0666667       13:-0.6  14:-0.0666667       15:-0.2   
2        11:0.2       12:-0.2  13:-0.866667         14:0.2       15:-0.6   
3  11:-0.333333  12:-0.333333  13:-0.866667         14:0.2  15:-0.733333   
4   11:0.333333  12:-0.466667  13:-0.733333    14:0.466667  15:-0.733333   

          feat16  feat17  
0   16:0.0666667     NaN  
1    16:0.466667     NaN  
2  16:-0.0666667     NaN  
3   16:0.0666667     NaN  
4   16:-0.466667     NaN

In [212]:
columns=['feat1','feat2','feat3','feat4','feat5','feat6','feat7','feat8','feat9','feat10','feat11','feat12','feat13','feat14','feat15','feat16']
# Removing the first two characters in every data attribute value in the test data set
for c in columns:
    test_data[c] = test_data[c].str.split(':').str[1]
test_data.head()

Alphabet      feat1       feat2      feat3      feat4      feat5  \
0        17  -0.466667  -0.0666667  -0.333333       -0.2  -0.733333   
1        18       -0.6         0.2  -0.466667       -0.2       -0.6   
2         6  -0.733333   -0.866667  -0.733333  -0.733333  -0.866667   
3         6  -0.733333   -0.466667  -0.466667       -0.6  -0.733333   
4        20  -0.333333   0.0666667       -0.2       -0.2       -0.6   

        feat6     feat7      feat8       feat9      feat10     feat11  \
0   0.0666667      -0.2  0.0666667        -0.2        -0.2       -0.2   
1   -0.333333  0.333333        0.2        -0.6  -0.0666667  -0.466667   
2   -0.333333  0.333333  -0.466667   -0.333333    0.333333        0.2   
3  -0.0666667       0.2  -0.866667  -0.0666667    0.733333  -0.333333   
4   -0.466667  0.733333  -0.466667        -0.2         0.6   0.333333   

      feat12     feat13      feat14     feat15      feat16  feat17  
0  0.0666667       -0.6   0.0666667  -0.466667   0.0666667     NaN  
1  0.0666667       -0.6  -0.0666667       -0.2    0.466667     NaN  
2       -0.2  -0.866667         0.2       -0.6  -0.0666667     NaN  
3  -0.333333  -0.866667         0.2  -0.733333   0.0666667     NaN  
4  -0.466667  -0.733333    0.466667  -0.733333   -0.466667     NaN

In [213]:
train_data.isnull().any()

Alphabet    False
feat1       False
feat2       False
feat3       False
feat4       False
feat5       False
feat6       False
feat7       False
feat8       False
feat9       False
feat10      False
feat11      False
feat12      False
feat13      False
feat14      False
feat15      False
feat16      False
feat17       True
dtype: bool

In [214]:
test_data.isnull().any()

Alphabet    False
feat1       False
feat2       False
feat3       False
feat4       False
feat5       False
feat6       False
feat7       False
feat8       False
feat9       False
feat10      False
feat11      False
feat12      False
feat13      False
feat14      False
feat15      False
feat16      False
feat17       True
dtype: bool

In [215]:
train_data=train_data.dropna(axis=1,how='all') #Dropping NaNs
train_data.head()

Alphabet       feat1       feat2      feat3       feat4      feat5  \
0        26   -0.733333   -0.466667  -0.466667        -0.6  -0.733333   
1        16   -0.466667  -0.0666667  -0.333333   -0.333333  -0.333333   
2        19  -0.0666667    0.333333  0.0666667  -0.0666667  -0.466667   
3         8   -0.466667         0.2  -0.333333  -0.0666667  -0.466667   
4         8        -0.2  -0.0666667  0.0666667   -0.333333  -0.466667   

        feat6       feat7      feat8       feat9      feat10      feat11  \
0  -0.0666667   0.0666667  -0.733333         0.2    0.466667  -0.0666667   
1   -0.333333         0.2       -0.2   -0.466667   0.0666667  -0.0666667   
2   0.0666667   0.0666667  -0.333333    0.333333    0.466667   -0.733333   
3  -0.0666667  -0.0666667   0.733333   -0.866667  -0.0666667        -0.2   
4  -0.0666667        -0.2       -0.6  -0.0666667    0.333333  -0.0666667   

       feat12     feat13     feat14      feat15      feat16  
0  -0.0666667  -0.866667  0.0666667   -0.333333        -0.2  
1         0.2  -0.733333        0.2  -0.0666667    0.333333  
2   0.0666667  -0.733333  -0.333333   -0.333333    0.333333  
3   0.0666667       -0.6  0.0666667          -1   0.0666667  
4         0.2       -0.6  0.0666667        -0.6  -0.0666667

In [216]:
train_data.head()

Alphabet       feat1       feat2      feat3       feat4      feat5  \
0        26   -0.733333   -0.466667  -0.466667        -0.6  -0.733333   
1        16   -0.466667  -0.0666667  -0.333333   -0.333333  -0.333333   
2        19  -0.0666667    0.333333  0.0666667  -0.0666667  -0.466667   
3         8   -0.466667         0.2  -0.333333  -0.0666667  -0.466667   
4         8        -0.2  -0.0666667  0.0666667   -0.333333  -0.466667   

        feat6       feat7      feat8       feat9      feat10      feat11  \
0  -0.0666667   0.0666667  -0.733333         0.2    0.466667  -0.0666667   
1   -0.333333         0.2       -0.2   -0.466667   0.0666667  -0.0666667   
2   0.0666667   0.0666667  -0.333333    0.333333    0.466667   -0.733333   
3  -0.0666667  -0.0666667   0.733333   -0.866667  -0.0666667        -0.2   
4  -0.0666667        -0.2       -0.6  -0.0666667    0.333333  -0.0666667   

       feat12     feat13     feat14      feat15      feat16  
0  -0.0666667  -0.866667  0.0666667   -0.333333        -0.2  
1         0.2  -0.733333        0.2  -0.0666667    0.333333  
2   0.0666667  -0.733333  -0.333333   -0.333333    0.333333  
3   0.0666667       -0.6  0.0666667          -1   0.0666667  
4         0.2       -0.6  0.0666667        -0.6  -0.0666667

In [217]:
test_data=test_data.dropna(axis=1,how='all') # Dropping NaNs

In [218]:
test_data.head()

Alphabet      feat1       feat2      feat3      feat4      feat5  \
0        17  -0.466667  -0.0666667  -0.333333       -0.2  -0.733333   
1        18       -0.6         0.2  -0.466667       -0.2       -0.6   
2         6  -0.733333   -0.866667  -0.733333  -0.733333  -0.866667   
3         6  -0.733333   -0.466667  -0.466667       -0.6  -0.733333   
4        20  -0.333333   0.0666667       -0.2       -0.2       -0.6   

        feat6     feat7      feat8       feat9      feat10     feat11  \
0   0.0666667      -0.2  0.0666667        -0.2        -0.2       -0.2   
1   -0.333333  0.333333        0.2        -0.6  -0.0666667  -0.466667   
2   -0.333333  0.333333  -0.466667   -0.333333    0.333333        0.2   
3  -0.0666667       0.2  -0.866667  -0.0666667    0.733333  -0.333333   
4   -0.466667  0.733333  -0.466667        -0.2         0.6   0.333333   

      feat12     feat13      feat14     feat15      feat16  
0  0.0666667       -0.6   0.0666667  -0.466667   0.0666667  
1  0.0666667       -0.6  -0.0666667       -0.2    0.466667  
2       -0.2  -0.866667         0.2       -0.6  -0.0666667  
3  -0.333333  -0.866667         0.2  -0.733333   0.0666667  
4  -0.466667  -0.733333    0.466667  -0.733333   -0.466667

In [219]:
train_data.isnull().any()

Alphabet    False
feat1       False
feat2       False
feat3       False
feat4       False
feat5       False
feat6       False
feat7       False
feat8       False
feat9       False
feat10      False
feat11      False
feat12      False
feat13      False
feat14      False
feat15      False
feat16      False
dtype: bool

In [220]:
test_data.isnull().any()

Alphabet    False
feat1       False
feat2       False
feat3       False
feat4       False
feat5       False
feat6       False
feat7       False
feat8       False
feat9       False
feat10      False
feat11      False
feat12      False
feat13      False
feat14      False
feat15      False
feat16      False
dtype: bool

In [221]:
#from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier

In [222]:
# Defining our training features and label
training_features=train_data.drop('Alphabet',axis=1)
train_label=train_data['Alphabet']

In [223]:
# Define our test data features and label
testing_features=test_data.drop('Alphabet',axis=1)
test_label=test_data['Alphabet']

In [224]:
X_train=training_features
y_train=train_label
X_test=testing_features
y_test=test_label

In [292]:
# By taking all the features, we have defined our max_depth respectively and iteratively ran the classifier to generate accuracies.
from sklearn.metrics import accuracy_score,classification_report,confusion_matrix
max_depth_range=[5,10,15,20,25]
scores=[]
for depth in max_depth_range:
    dtree=DecisionTreeClassifier(max_depth=depth,criterion='entropy')
    estimator=dtree.fit(X_train,y_train)
    pred=dtree.predict(X_test)
    score=accuracy_score(pred,y_test)
    scores.append(score)
    
print(scores)
#print(classification_report(pred,y_test))

[0.5014, 0.7992, 0.8738, 0.8714, 0.8748]


In [293]:
# By taking all the features, we have defined our max_depth respectively and iteratively ran the Gini criterion classifier to generate accuracies.

max_depth_range=[5,10,15,20,25]
scores=[]
for depth in max_depth_range:
    dtree=DecisionTreeClassifier(max_depth=depth,criterion='gini')
    estimator=dtree.fit(X_train,y_train)
    pred_g=dtree.predict(X_test)
    score=accuracy_score(pred_g,y_test)
    scores.append(score)
    
print(scores)
#print(classification_report(pred_g,y_test))

[0.3694, 0.7134, 0.8324, 0.8642, 0.8764]


In [233]:
# Classification report generates all the required metrics. And we have selectively ran the algorithm to generate reports for each max_depth respectively.
from sklearn.metrics import classification_report,confusion_matrix
print(classification_report(y_test,pred))

              precision    recall  f1-score   support

           1       0.95      0.95      0.95       195
           2       0.79      0.79      0.79       199
           3       0.90      0.86      0.88       182
           4       0.84      0.86      0.85       207
           5       0.87      0.89      0.88       203
           6       0.90      0.87      0.88       210
           7       0.88      0.81      0.84       226
           8       0.78      0.83      0.80       196
           9       0.92      0.90      0.91       188
          10       0.88      0.89      0.88       172
          11       0.80      0.83      0.82       167
          12       0.92      0.94      0.93       201
          13       0.93      0.91      0.92       185
          14       0.89      0.90      0.90       178
          15       0.83      0.83      0.83       179
          16       0.88      0.89      0.88       206
          17       0.84      0.81      0.83       189
          18       0.77    

In [282]:
dtree=DecisionTreeClassifier(max_depth=5,criterion='entropy')
start_time_e5=time.time()
estimator=dtree.fit(X_train,y_train)
end_time_e5=time.time()
pred1=dtree.predict(X_test)
score=accuracy_score(pred1,y_test)
print(f"Training time: {end_time_e5-start_time_e5}s")
print(score)
print(classification_report(y_test,pred1))

Training time: 0.05827617645263672s
0.5014
              precision    recall  f1-score   support

           1       0.75      0.75      0.75       195
           2       0.31      0.70      0.43       199
           3       0.41      0.64      0.50       182
           4       0.54      0.61      0.57       207
           5       0.46      0.41      0.43       203
           6       0.54      0.65      0.59       210
           7       0.00      0.00      0.00       226
           8       0.57      0.29      0.38       196
           9       0.60      0.69      0.64       188
          10       0.96      0.57      0.72       172
          11       0.76      0.17      0.28       167
          12       0.80      0.63      0.71       201
          13       0.98      0.67      0.80       185
          14       0.83      0.61      0.71       178
          15       0.25      0.53      0.34       179
          16       0.78      0.67      0.72       206
          17       0.19      0.28     

/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/sklearn/metrics/classification.py:1437: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)


In [281]:
dtree=DecisionTreeClassifier(max_depth=10,criterion='entropy')
start_time_e10=time.time()
estimator=dtree.fit(X_train,y_train)
end_time_e10=time.time()
pred2=dtree.predict(X_test)
score=accuracy_score(pred2,y_test)
print(f"Training time: {end_time_e10-start_time_e10}s")
print(score)
print(classification_report(y_test,pred2))

Training time: 0.07447195053100586s
0.7986
              precision    recall  f1-score   support

           1       0.94      0.90      0.92       195
           2       0.68      0.80      0.74       199
           3       0.80      0.80      0.80       182
           4       0.72      0.83      0.77       207
           5       0.69      0.83      0.76       203
           6       0.73      0.84      0.78       210
           7       0.73      0.67      0.70       226
           8       0.65      0.74      0.69       196
           9       0.94      0.81      0.87       188
          10       0.91      0.86      0.89       172
          11       0.63      0.58      0.60       167
          12       0.96      0.83      0.89       201
          13       0.87      0.92      0.90       185
          14       0.88      0.87      0.87       178
          15       0.76      0.73      0.74       179
          16       0.93      0.84      0.88       206
          17       0.79      0.74     

In [287]:
dtree=DecisionTreeClassifier(max_depth=15,criterion='entropy')
start_time_e15=time.time()
estimator=dtree.fit(X_train,y_train)
end_time_e15=time.time()
pred3=dtree.predict(X_test)
score=accuracy_score(pred3,y_test) 
print(f"Training time: {end_time_e15-start_time_e15}s")
print(score)
print(classification_report(y_test,pred3))

Training time: 0.09489202499389648s
0.875
              precision    recall  f1-score   support

           1       0.93      0.93      0.93       195
           2       0.81      0.84      0.83       199
           3       0.93      0.90      0.91       182
           4       0.85      0.89      0.87       207
           5       0.83      0.83      0.83       203
           6       0.83      0.86      0.85       210
           7       0.88      0.81      0.84       226
           8       0.85      0.80      0.82       196
           9       0.90      0.88      0.89       188
          10       0.85      0.89      0.87       172
          11       0.83      0.87      0.85       167
          12       0.92      0.93      0.93       201
          13       0.89      0.93      0.91       185
          14       0.90      0.92      0.91       178
          15       0.88      0.84      0.86       179
          16       0.89      0.90      0.90       206
          17       0.89      0.86      

In [278]:
dtree=DecisionTreeClassifier(max_depth=20,criterion='entropy')
start_time_e20=time.time()
estimator=dtree.fit(X_train,y_train)
end_time_e20=time.time()
pred4=dtree.predict(X_test)
score=accuracy_score(pred4,y_test)
print(f"Training time: {end_time_e20-start_time_e20}s")
print(score)
print(classification_report(y_test,pred4))

Training time: 0.09274411201477051s
0.8706
              precision    recall  f1-score   support

           1       0.94      0.93      0.93       195
           2       0.83      0.82      0.82       199
           3       0.94      0.89      0.92       182
           4       0.84      0.87      0.86       207
           5       0.87      0.83      0.85       203
           6       0.84      0.84      0.84       210
           7       0.83      0.81      0.82       226
           8       0.84      0.79      0.81       196
           9       0.91      0.87      0.89       188
          10       0.86      0.92      0.89       172
          11       0.81      0.86      0.83       167
          12       0.93      0.93      0.93       201
          13       0.91      0.93      0.92       185
          14       0.88      0.93      0.90       178
          15       0.88      0.86      0.87       179
          16       0.91      0.90      0.91       206
          17       0.82      0.87     

In [277]:
dtree=DecisionTreeClassifier(max_depth=25,criterion='entropy')
start_time_e25=time.time()
estimator=dtree.fit(X_train,y_train)
end_time_e25=time.time()
pred5=dtree.predict(X_test)
score=accuracy_score(pred5,y_test)
print(f"Training time: {end_time_e25-start_time_e25}s")
print(score)
print(classification_report(y_test,pred5))

Training time: 0.09412765502929688s
0.8728
              precision    recall  f1-score   support

           1       0.92      0.93      0.93       195
           2       0.82      0.82      0.82       199
           3       0.91      0.90      0.91       182
           4       0.86      0.88      0.87       207
           5       0.83      0.84      0.84       203
           6       0.85      0.84      0.85       210
           7       0.87      0.82      0.84       226
           8       0.83      0.82      0.82       196
           9       0.92      0.89      0.91       188
          10       0.88      0.90      0.89       172
          11       0.82      0.85      0.83       167
          12       0.94      0.94      0.94       201
          13       0.91      0.94      0.92       185
          14       0.90      0.91      0.90       178
          15       0.87      0.85      0.86       179
          16       0.92      0.91      0.91       206
          17       0.85      0.84     

In [276]:
dtree=DecisionTreeClassifier(max_depth=5,criterion='gini')
start_time_g5=time.time()
estimator=dtree.fit(X_train,y_train)
end_time_g5=time.time()
pred_1=dtree.predict(X_test)
score=accuracy_score(pred_1,y_test)
print(f"Training time: {end_time_g5-start_time_g5}s")
print(score)
print(classification_report(y_test,pred_1))

Training time: 0.0548248291015625s
0.3694
              precision    recall  f1-score   support

           1       1.00      0.74      0.85       195
           2       0.00      0.00      0.00       199
           3       0.00      0.00      0.00       182
           4       0.20      0.85      0.32       207
           5       0.98      0.28      0.44       203
           6       0.00      0.00      0.00       210
           7       0.00      0.00      0.00       226
           8       0.00      0.00      0.00       196
           9       0.49      0.77      0.60       188
          10       0.85      0.74      0.79       172
          11       0.00      0.00      0.00       167
          12       0.99      0.69      0.82       201
          13       0.93      0.69      0.79       185
          14       0.61      0.08      0.14       178
          15       0.00      0.00      0.00       179
          16       0.61      0.54      0.57       206
          17       0.10      0.83      

/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/sklearn/metrics/classification.py:1437: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)


In [275]:
dtree=DecisionTreeClassifier(max_depth=10,criterion='gini')
start_time_g10=time.time()
estimator=dtree.fit(X_train,y_train)
end_time_g10=time.time()
pred_2=dtree.predict(X_test)
score=accuracy_score(pred_2,y_test)
print(f"Training time: {end_time_g10-start_time_g10}s")
print(score)
print(classification_report(y_test,pred_2))

Training time: 0.07276010513305664s
0.7142
              precision    recall  f1-score   support

           1       0.97      0.78      0.87       195
           2       0.66      0.62      0.64       199
           3       0.38      0.79      0.52       182
           4       0.71      0.77      0.74       207
           5       0.69      0.62      0.65       203
           6       0.74      0.76      0.75       210
           7       0.81      0.55      0.66       226
           8       0.32      0.44      0.37       196
           9       0.90      0.81      0.85       188
          10       0.97      0.80      0.88       172
          11       0.52      0.71      0.60       167
          12       0.98      0.78      0.87       201
          13       0.91      0.85      0.88       185
          14       0.91      0.84      0.87       178
          15       0.39      0.75      0.52       179
          16       0.93      0.70      0.80       206
          17       0.52      0.64     

In [274]:
dtree=DecisionTreeClassifier(max_depth=15,criterion='gini')
start_time_g15=time.time()
estimator=dtree.fit(X_train,y_train)
end_time_g15=time.time()
pred_3=dtree.predict(X_test)
score=accuracy_score(pred_3,y_test)
print(f"Training time: {end_time_g15-start_time_g15}s")
print(score)
print(classification_report(y_test,pred_3))

Training time: 0.08896183967590332s
0.8328
              precision    recall  f1-score   support

           1       0.84      0.91      0.87       195
           2       0.58      0.80      0.68       199
           3       0.88      0.84      0.86       182
           4       0.84      0.83      0.83       207
           5       0.85      0.83      0.84       203
           6       0.87      0.82      0.85       210
           7       0.63      0.76      0.69       226
           8       0.84      0.70      0.76       196
           9       0.91      0.88      0.89       188
          10       0.85      0.88      0.87       172
          11       0.64      0.85      0.73       167
          12       0.93      0.92      0.92       201
          13       0.95      0.88      0.91       185
          14       0.91      0.85      0.88       178
          15       0.75      0.70      0.72       179
          16       0.91      0.84      0.87       206
          17       0.85      0.73     

In [273]:
dtree=DecisionTreeClassifier(max_depth=20,criterion='gini')
start_time_g20=time.time()
estimator=dtree.fit(X_train,y_train)
end_time_g20=time.time()
pred_4=dtree.predict(X_test)
score=accuracy_score(pred_4,y_test)  
print(f"Training time: {end_time_g20-start_time_g20}s")
print(score)
print(classification_report(y_test,pred_4))

Training time: 0.09591984748840332s
0.8678
              precision    recall  f1-score   support

           1       0.93      0.95      0.94       195
           2       0.73      0.79      0.76       199
           3       0.87      0.85      0.86       182
           4       0.88      0.86      0.87       207
           5       0.86      0.86      0.86       203
           6       0.90      0.86      0.88       210
           7       0.85      0.81      0.83       226
           8       0.82      0.80      0.81       196
           9       0.93      0.91      0.92       188
          10       0.92      0.89      0.90       172
          11       0.83      0.82      0.83       167
          12       0.92      0.94      0.93       201
          13       0.81      0.92      0.87       185
          14       0.91      0.87      0.89       178
          15       0.84      0.80      0.82       179
          16       0.85      0.88      0.86       206
          17       0.84      0.83     

In [272]:
dtree=DecisionTreeClassifier(max_depth=25,criterion='gini')
start_time_g25=time.time()
estimator=dtree.fit(X_train,y_train)
end_time_g25=time.time()
pred_5=dtree.predict(X_test)
score=accuracy_score(pred_5,y_test)  
print(f"Training time: {end_time_g25-start_time_g25}s")
print(score)
print(classification_report(y_test,pred_5))

Training time: 0.17548084259033203s
0.8756
              precision    recall  f1-score   support

           1       0.93      0.94      0.93       195
           2       0.82      0.77      0.80       199
           3       0.88      0.86      0.87       182
           4       0.86      0.88      0.87       207
           5       0.87      0.86      0.86       203
           6       0.91      0.86      0.88       210
           7       0.88      0.82      0.85       226
           8       0.80      0.82      0.81       196
           9       0.92      0.90      0.91       188
          10       0.90      0.89      0.89       172
          11       0.80      0.84      0.82       167
          12       0.93      0.93      0.93       201
          13       0.90      0.92      0.91       185
          14       0.91      0.89      0.90       178
          15       0.83      0.85      0.84       179
          16       0.88      0.91      0.90       206
          17       0.88      0.84     

# We didn't change the parameters or tune them infact, and the dataset has been taken without any validation and this algorithm generates around 88 % accuracy